kafka.go

```diff
@@ -14,148 +14,270 @@
+		defer func() {
+			_ = reader.Close()
+		}()
 
-	return err
+		if err := reader.SetOffset(kafka.LastOffset); err != nil {
+			logrus.Error("unable to set offset. error : ", err)
+			return
+		}
+
+		for {
+			consumerMessage, err := reader.ReadMessage(subscribeBuilder.ctx)
+			if err != nil {
+				logrus.Error("unable to subscribe topic from kafka. error : ", err)
+				return
+			}
+			go client.processMessage(consumerMessage)
+		}
+	}()
 }
 
-// PublishEventAsync publish to a Kafka topic asynchronously
-func (client *KafkaClient) PublishEventAsync(event *Event) {
-	event.Realm = client.realm
-	event.Time = time.Now().UTC()
+// registerCallback add callback to map with topic and eventName as a key
+func (client *KafkaClient) registerCallback(topic, eventName string, callback func(event *Event, err error)) (
+	isRegistered bool) {
 
-	client.asyncProducer.Input() <- &sarama.ProducerMessage{
-		Timestamp: time.Now().UTC(),
-		Value: &eventEntry{
-			Event: *event,
-		},
-		Topic: event.Topic,
+	if innerMap, isExists := client.subscribeMap.Load(topic); isExists {
+		if _, isEventName := innerMap.(*sync.Map).Load(eventName); isEventName {
+			return true
+		}
 	}
+
+	var callbackMap = &sync.Map{}
+	callbackMap.Store(eventName, callback)
+	client.subscribeMap.Store(topic, callbackMap)
+	return false
 }
 
-// configureKafka configure kafka
-func configureKafka(kafkaConfig *KafkaConfig) (config *sarama.Config) {
-	config = sarama.NewConfig()
-	if kafkaConfig.DialTimeout != 0 {
-		config.Net.DialTimeout = kafkaConfig.DialTimeout
-	}
-	if kafkaConfig.ReadTimeout != 0 {
-		config.Net.ReadTimeout = kafkaConfig.ReadTimeout
+// processMessage process a message from kafka
+func (client *KafkaClient) processMessage(message kafka.Message) {
+	event, err := unmarshal(message)
+	if err != nil {
+		logrus.Error("unable to unmarshal message from subscribe in kafka. error : ", err)
+		return
 	}
-	if kafkaConfig.WriteTimeout != 0 {
-		config.Net.WriteTimeout = kafkaConfig.WriteTimeout
+	client.runCallback(event, message)
+}
+
+// unmarshal unmarshal received message into event struct
+func unmarshal(message kafka.Message) (*Event, error) {
+	var event Event
+	err := json.Unmarshal(message.Value, &event)
+	if err != nil {
+		return &Event{}, err
 	}
-	if kafkaConfig.MetadataRetryMax != 0 {
-		config.Metadata.Retry.Max = kafkaConfig.MetadataRetryMax
+	return &event, nil
+}
+
+// runCallback run callback function when receive an event
+func (client *KafkaClient) runCallback(event *Event, consumerMessage kafka.Message) {
+	innerMap, ok := client.subscribeMap.Load(consumerMessage.Topic)
+	if !ok {
+		logrus.Warnf("callback not found for topic : %s", consumerMessage.Topic)
+		return
 	}
-	if kafkaConfig.MetadataRetryBackoff != 0 {
-		config.Metadata.Retry.Backoff = kafkaConfig.MetadataRetryBackoff
+	callback, ok := innerMap.(*sync.Map).Load(event.EventName)
+	if !ok {
+		logrus.Warnf("callback not found for topic : %s, event name : %s", consumerMessage.Topic,
```

Comment:
it's not that callback is not found, it has wrong type

Reply:
Wrong type would make a panic and `ok` is a status of existing key
So, I think we need to add `recover` to catch the panic.
you can check this example https://play.golang.org/p/-brKdicRmqr

Reply:
i know, I'm talking about the error message

Reply:
oh I'm sorry, I'm not clear enough about it, do you mean the log type is wrong ?

Reply:
I realized there's even more dangerous problem.

So at this point in the code, the innerMap is found in the map, but the type is not checked.
If the type of `innerMap` is not `*sync.Map`, it should fail with error "wrong type".

The right way here, is to have type-casting separate than the `Load()`. Then it can be queried using `Load`. If it's not found, then have "not found" error.

Then, the callback itself need to be type-checked by type-casting before being called in the return statement. This also need a "wrong type" error.

That means you need to separate type-casting and calling the callback.

if you need refresher on type-casting and type assertion:
https://tour.golang.org/methods/15
line 11 and 14

Reply:
ah I see, I think its a safest way instead of recover the panic because we dont know the panic caused by error type-casting or another thing. Thanks